tmol/score/rama/rama_energy_term.py

```diff
@@ -0,0 +1,140 @@
+
+from tmol.score.energy_term import EnergyTerm
+from .params import RamaBlockTypeParams, RamaPackedBlockTypesParams, RamaParamResolver
+from .rama_whole_pose_module import RamaWholePoseScoringModule
+from tmol.database import ParameterDatabase
+
+from tmol.chemical.restypes import RefinedResidueType, uaid_t
+from tmol.pose.packed_block_types import PackedBlockTypes
+from tmol.pose.pose_stack import PoseStack
+
+
+class RamaEnergyTerm(EnergyTerm):
+    device: torch.device
+    param_resolver: RamaParamResolver
+    tables: Tensor[torch.float32][:, :, :]
+    table_params: Tensor[torch.float32][:, :, :]
+
+    def __init__(self, param_db: ParameterDatabase, device: torch.device):
+        super(RamaEnergyTerm, self).__init__(param_db=param_db, device=device)
+        self.device = device
+
+        self.param_resolver = RamaParamResolver.from_database(
+            param_db.scoring.rama, device
+        )
+
+        self.tables = self.param_resolver.rama_params.tables
+        self.table_params = torch.cat(
+            (
+                self.param_resolver.rama_params.bbstarts.to(torch.float32),
+                self.param_resolver.rama_params.bbsteps.to(torch.float32),
+            ),
+            dim=1,
+        )
+
+    @classmethod
+    def score_types(cls):
+        import tmol.score.terms.rama_creator
+
+        return tmol.score.terms.rama_creator.RamaTermCreator.score_types()
+
+    def n_bodies(self):
+        return 2
+
+    def setup_block_type(self, block_type: RefinedResidueType):
+        super(RamaEnergyTerm, self).setup_block_type(block_type)
+        if hasattr(block_type, "rama_params"):
+            return
+
+        rname = block_type.name
+        lookups = numpy.array([[rname, "_"], [rname, "PRO"]], dtype=object)
+        table_inds = self.param_resolver.rama_lookup.index.get_indexer(lookups)
+        table_inds = self.param_resolver.rama_lookup.iloc[table_inds, :][
+            "table_id"
+        ].values
+
+        rama_torsion_atoms = numpy.full((2, 4), -1, dtype=uaid_t)
+
+        def uaids_to_np(tor_uaids):
+            uaids = numpy.full((4,), -1, dtype=uaid_t)
+            for i in range(4):
+                uaids[i] = numpy.array(tor_uaids[i], dtype=uaid_t)
+            return uaids
+
+        if table_inds[0] != -1 or table_inds[1] != -1:
+            for i, tor in enumerate(["phi", "psi"]):
+                if tor in block_type.torsion_to_uaids:
+                    rama_torsion_atoms[i] = uaids_to_np(
+                        block_type.torsion_to_uaids[tor]
+                    )
+        rama_torsion_atoms = rama_torsion_atoms.reshape(8)
+
+        upper_conn_id = numpy.full((1,), -1, dtype=numpy.int32)
+        if "up" in block_type.connection_to_cidx:
+            upper_conn_id[0] = block_type.connection_to_cidx["up"]
+        is_pro = numpy.full((1,), 0, dtype=numpy.int32)
+
+        # TO DO: Better logic here to handle proline variants
+        if block_type.name == "PRO":
```

Comment:
we need to think about this (and it arises much worse in fa_dun).  I think there needs to be two parts: 1) what does this score term consider a proline?, and 2) how are patched residue types specified?  The answer to 1 is probably different for different score terms.

This is fine for now though.